Top level function binding + STG AST

stormeuh opened the pull request.

stormeuh committed:
Initial implementation of STG syntax

stormeuh committed:
Added top level function declaration support

stormeuh committed:
First implementation of STG pretty printing

Gertjan423 commented on Backend/STGTypes.hs:
```diff
@@ -0,0 +1,163 @@
+module Backend.STGTypes where
```
I don't think we should just put these in the existing Backend directory. We should make a clear distinction between the system f intermediate system, and the stg backend language.

Gertjan423 commented on Backend/STGTypes.hs:
```diff
@@ -0,0 +1,163 @@
+
+-- Variable
+newtype SVar = SVar {svar_name :: Name}
+newtype SCon = SCon {scon_clos :: SVar}
```
Can you write a comment here? I'm confused what this SCon represents?

Gertjan423 commented on Backend/STGTypes.hs:
```diff
@@ -0,0 +1,163 @@
+
+-- Variable binding (non-empty list of bindings)
+data SBinds = SBinds SVar SLForm SBinds
+            | SBind  SVar SLForm
```
While I can see the intent, it seems a bit clumsy to me, to duplicate this constructor. A possible alternative 
`SBind Svar SLForm (Maybe SBinds)`?

Gertjan423 commented on Backend/STGTypes.hs:
```diff
@@ -0,0 +1,163 @@
+-- * Expressions and atoms
+-- ----------------------------------------------------------------------------
+
+-- Expression
```
Maybe add a reference to where you found this definition of STG expressions, so a reader knows where to find more information?

Gertjan423 commented on Frontend/HsTypeChecker.hs:
```diff
@@ -778,6 +783,27 @@ extendCtxKindAnnotatedTysM ann_as = extendCtxTysM as (map kindOf as)
+-- * Function Declaration Elaboration
+-- ------------------------------------------------------------------------------
+
+-- | 
```
This comment could be slightly more informative :)

Gertjan423 commented on Frontend/HsTypeChecker.hs:
```diff
@@ -990,6 +1016,13 @@ elabProgram theory (PgmData data_decl pgm) = do
   let fc_program = FcPgmDataDecl fc_data_decl fc_pgm
   return (fc_program, ty, final_theory)
 
+-- Elaborate a function declaration
```
Note that we write function comments using `-- | `, in order to be haddock compatible

Gertjan423 commented on Tests/Prelude.hs:
```diff
@@ -0,0 +1,730 @@
+-- General/Miscellaneous functions
```
Do include a source here, as you didn't write this code yourself

Gertjan423 commented on Utils/PrettyPrint.hs:
```diff
@@ -85,6 +86,10 @@ cat  = liftListOp P.cat
 fsep = liftListOp P.fsep
 fcat = liftListOp P.fcat
 
+hcatmap, hsepmap :: PrettyPrint a => (a -> Doc) -> [a] -> Doc
```
Include a comment, explaining this function

Gertjan423 reviewed (changes requested):
Looks good! Some small comments though.

stormeuh committed:
Created directories for Optimizer, Interpreter and Compiler

Moved and renamed Backend.FcEvaluator to Backend.Interpreter.FcEvaluator
Moved and renamed Backend.FcTypeChecker and FcTypes to
Optimizer.FcTypeChecker and Optimizer.FcTypes resp.

stormeuh committed:
made documentation haddock-compliant, changed SBinds definition

stormeuh committed:
added documentation
